modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java

```diff
@@ -0,0 +1,327 @@
+    }
+
+    private CompletableFuture<?> onIndexDrop(DropIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            int indexId = parameters.indexId();
+
+            int partitions = getPartitionCountFromCatalog(indexId, parameters.catalogVersion() - 1);
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
+
+            return metaStorageManager.invoke(
+                    exists(startBuildIndexKey),
+                    concat(
+                            List.of(remove(startBuildIndexKey)),
+                            removePartitionBuildIndexOperations(indexId, partitions)
+                    ),
+                    List.of(noop())
+            );
+        });
+    }
+
+    private CompletableFuture<?> onIndexAvailable(MakeIndexAvailableEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            ByteArray startBuildIndexKey = startBuildIndexKey(parameters.indexId());
+
+            return metaStorageManager.invoke(exists(startBuildIndexKey), remove(startBuildIndexKey), noop());
+        });
+    }
+
+    private CompletableFuture<?> onUpdatePartitionBuildIndexKey(WatchEvent event) {
+        return inBusyLockAsync(busyLock, () -> {
+            if (!event.single()) {
+                // We don't need to handle keys on index creation or deletion.
+                return completedFuture(null);
+            }
+
+            Entry entry = event.entryEvent().newEntry();
+
+            if (!entry.tombstone()) {
+                // In case an index was created when there was only one partition.
+                return completedFuture(null);
+            }
+
+            String partitionBuildIndexKey = new String(entry.key(), UTF_8);
+
+            int indexId = parseIndexIdFromPartitionBuildIndexKey(partitionBuildIndexKey);
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
+
+            long metastoreRevision = entry.revision();
+
+            if (isRemainingPartitionBuildIndexKeys(indexId, metastoreRevision)
+                    || isMetastoreKeyAbsent(startBuildIndexKey, metastoreRevision)) {
+                return completedFuture(null);
+            }
+
+            // We can use the latest version of the catalog since we are on the metastore thread.
+            CatalogIndexDescriptor indexDescriptor = getIndexDescriptorStrict(indexId, catalogManager.latestCatalogVersion());
+
+            // We will not wait for the command to be executed, since we will then find ourselves in a dead lock since we will not be able
+            // to free the metastore thread.
+            catalogManager.execute(buildMakeIndexAvailableCommand(indexDescriptor));
+
+            return completedFuture(null);
+        });
+    }
+
+    private void onIndexBuildForPartition(int indexId, int partitionId) {
+        inBusyLock(busyLock, () -> {
+            ByteArray partitionBuildIndexKey = partitionBuildIndexKey(indexId, partitionId);
```

Comment:
Does a node know somehow that it finished building an index for a partition, is this persisted somewhere else? Imagine a scenario when an index gets built for partition 1, but before the key is removed from the MetaStorage the server gets restarted. Will it be able to see on recovery that the index has already been built for partition 1? I realize this should not be implemented in this PR (it's not about recovery), but I'm still curious.

Reply:
I indicated this problem in the recovery ticket [IGNITE-20637](https://issues.apache.org/jira/browse/IGNITE-20637) and we will need to carefully solve it.
At the moment, in my understanding, during recovery we will need to wait until the replication groups with the replication log applied are raised, and only after that will it be possible to look into the index storage and find out whether the index has been built, if so, then it will be possible to delete the index construction key for the partition.